### ISSUE TITLE
Image component delaying resource request due to requestIdleCallback being heavily throttled in newer Chrome

### ISSUE THREAD

**Issue by sidhu663** [necolas/react-native-web#759]
**Do you want to request a *feature* or report a *bug*?**
Bug / regression

**What is the current behavior?**
Image component uses requestIdleCallback to schedule the resource request. 
On newer Chrome, rIC is heavily throttled when a user scrolls, adding approximately 500ms - 2000ms delay for just fetching the image.

**If the current behavior is a bug, please provide the steps to reproduce and
if a minimal demo of the problem via Glitch or similar (template:
https://glitch.com/edit/#!/react-native-web-playground).**

jsFiddle where the issue can be reproduced: https://jsfiddle.net/zedt5szm/2/

When no user scroll:
1. Click start
2. Do not scroll
3. Observe that the time waited before executing the rIC is ~1ms

When user scrolls at least once:
1. Click start
2. Scroll once or twice before work is completed (within 2s of clicking start)
3. Observe that the time waited before executing the rIC is > 1000ms

When user continually scrolls while work is happening (this is the case with a virtualized scroller)
1. Click start
2. Keep scrolling up and down for about 3s (longer than the time work is being done)
3. Observe that the time waited before executing the rIC is ~2000ms after the most recent scroll event.

This results in a consistent 2s delay for all image loads when scrolling

**What is the expected behavior?**
Ideally, we would want the image request to be kicked off without a second + delay.

I've not managed to confirm this but according to this discussion: https://github.com/aFarkas/lazysizes/issues/434#issuecomment-336057405 , Chrome on user scroll heavily throttles rIC and defers image decoding.

Given that this usage of rIC is presumably intended to defer decoding, this might not be necessary.

I've also confirmed that on older Chrome (56), rIC fires without the throttling.

**Environment (include versions). Did this work in previous versions?**

* OS: macOS 10.13.2
* Device: 2013 Macbook Pro
* Browser: Chrome 63
* React Native for Web (version): 0.1.14
* React (version): 16.00

**Discussion**
I'm not sure what the best solution to this is.

We could potentially:
1.  Replace rIC with requestAnimationFrame to prevent this delay.
2. Add option to disable this delay onto the Image component

**Comment by necolas** [comment]
We talked about this at length while I was at Twitter. Why would you want to load the main thread with extra work while you're scrolling through an infinite list? And have you looked at what other browsers do? I'm not sure rAF is the way to go, as that's for high-priority tasks.

**Comment by sidhu663** [comment]
The issue is that even if there is no JS to be done, there is a minimum of 2s delay before the resource is fetched. That's 2s+ that the main thread is not busy, yet the image is not requested. It makes for a poor user experience with that much latency. All other browsers (including older Chrome) do not add this additional throttling when the user scrolls.

Furthermore, if we are to presume that Chrome will defer image decoding in addition to this rIC throttling on scroll, we are not actually loading the main thread with any significant extra work.

I'm not sure if rAF is the way to go but the current delay is very problematic. 

An option to make it high priority would get around the immediate issue. 
Another option is to add a 'max delay' of sorts but whatever we pick as that delay will essentially be the minimum delay that Chrome users experience as long as it's less than the 2s built in throttled rIC.
We could also make use of https://www.chromestatus.com/feature/4897260684967936 (<image decode="async" />) in ImageLoader and this will get better as browsers support it.

Fundamentally, I don't think rIC makes any guarantees on when it executes which makes it undesirable to use when all images are being so massively delayed.

Happy to put up a patch if we can agree on a way to mitigate this

**Comment by necolas** [comment]
> All other browsers (including older Chrome) do not add this additional throttling when the user scrolls.

Perhaps something to bring up with the Chrome team

**Comment by sidhu663** [comment]
> Perhaps something to bring up with the Chrome team

We can do that but they're not really wrong in their implementation and as you said there are no guarantees on when rIC executes.

Are there any concerns with adding an option to schedule the image fetch with higher priority? This is the current experience: 

![slow load gif](https://user-images.githubusercontent.com/9156874/34702317-f19eee9c-f4a1-11e7-9af8-0a9016010e8f.gif)

There's no way to reduce that delay currently for what is essentially the largest chunk of users.

**Comment by necolas** [comment]
> We can do that but they're not really wrong in their implementation…

Well, no other browsers are doing this.

Why don't you poke around and try to find a solution, ideally one that doesn't alter the API or make image loading tasks compete with high priority work.

**Comment by necolas** [comment]
If you look into an opt-in/out mechanism for different image loading strategies, the `source` object seems like a good place to hide that away somewhat. React Native has a few (undocumented) platform-specific options in [there](https://github.com/facebook/react-native/blob/master/Libraries/Image/ImageSourcePropType.js). That would give you per-image control over priority, which seems useful (e.g., avatars before content?)

And I'm interested to see how you think `decode=async` might work - that's not something I'd heard about before, but good platform APIs are worth adopting 👍 

This also reminds me to add an `Image` benchmark…er, with scrolling

**Comment by comp615** [comment]
Linking to other discussion of similar issues:
https://github.com/whatwg/html/issues/1920

**Comment by sidhu663** [comment]
> And I'm interested to see how you think decode=async might work - that's not something I'd heard about before, but good platform APIs are worth adopting 👍

Assuming the intent with this being in rIC is to defer decoding of images while the main thread might be busy, `decode=async` should be able to replace rIC for this use case. 

If we're going to add priority to source, I'm inclined to follow the api of the decode option so:
```javascript
decode: oneOf(['async', 'sync'])
```

with sync using rAF (or rIC with timeout) and async using rIC such that when the decode support is available, it can be a straight forward transition.

**Comment by comp615** [comment]
@sidhu663 but how would you detect decode=async and know if you should rIC or not? Would that require a browser support map? I'm not sure that's something that you can feature detect well, but I haven't looked too much

**Comment by sidhu663** [comment]
It's not entirely clear yet how the feature detection will work but I'm presuming (optimistically) that's something that will be doable. 

That said, the feature detection aspect doesn't affect the choice of API.
In either `<Image decode="async" />` or `<Image source={{decode: 'async'}} />`, feature detection has to be available or there's no way to move forward to the native implementation regardless of our API choice.

**Comment by comp615** [comment]
I see. I think I agree then from the API perspective that matching the browser vendors suggestions makes sense. As you mention, I'm in favor of directly on the component since that seems to map best in my mind.

**Comment by necolas** [comment]
I looked into this a little. It seems that setting a rIC `timeout` breaks through whatever queueing Chrome is applying. So adding a default timeout or adding `timeout` to source seems like a good start.

**Comment by sidhu663** [comment]
The problem with rIC `timeout` is that it's effectively just a `setTimeout` during scrolls that exceed the timeout so I don't think a default is a good idea.

Adding a timeout option to source will help us resolve the issue in the interim (We can check the user agent and set a timeout on Chrome) but I don't think it will translate well if / when we introduce the native "decode=async" behavior. 

Presuming that decode/async is supported in the future, we would then have a timeout option that no longer makes sense because the deferring is handled by the browser rather than rIC.

If you think that's not an issue, I can put up a branch adding a timeout field to source, but my preference is to support a decode field that accepts sync/async and use rAF/rIC. The semantics of that field can translate directly to the native decode=sync/async

**Comment by necolas** [comment]
Image defers loading, not just decoding.

**Comment by sidhu663** [comment]
To be clear, I don't mean do UA sniffing in rnw, but in twitter code base.

> Your problem seemed to be with Chrome delaying rIC by 2s after a scroll, so I'm not sure why a shorter timeout is unacceptable.

What we want is to execute immediately *after* work is done (such as completing a scroll and main thread becoming idle) but adding a timeout of say `rIC(f, {timeout:500})` would execute 500ms after the Image is mounted rather than when the thread is actually idle (such as when the user is engaging in a scroll longer than 500ms in duration)

> And when would you want to sync decode if async is available?

That's a fair point but the option exists for some reason in the native implementation so I presume there's use cases I'm not considering

> Image defers loading, not just decoding.

I had presumed that the intent was to defer decoding since initiating the request doesn't generally lead to jank but if we want to keep the deferring of the request as well then I'll put up a patch with a deferTimeout option since that seems the best way forward

**Comment by necolas** [comment]
> What we want is to execute immediately after work is done…but adding a timeout…would execute 500ms after the Image is mounted rather than when the thread is actually idle

I'm familiar with rIC :). And hasn't Chrome decided that rIC won't actually run as soon as the thread is idle, e.g., this 2s delay after scroll completion? There is no _requirement_ to use rIC without a timeout. It's reasonable to consider providing a default threshold.

In the case of a slow scroll, it seems like an image will either move offscreen and be unmounted or it will have been on screen for `>n` ms and should probably start displaying itself. Is the main thread pretty busy during a slow scroll in Twitter Lite?

> the option exists for some reason in the native implementation

That's OK. There are lots of platform APIs that you can't access via the React Native components and APIs.

**Comment by sidhu663** [comment]
> I'm familiar with rIC :)

Sorry, just trying to be as clear as possible :)

> In the case of a slow scroll, it seems like an image will either move offscreen and be unmounted or it will have been on screen for >n ms and should probably start displaying itself.

Fair point, seems like we're agreed on a deferTimeout then.

> Is the main thread pretty busy during a slow scroll in Twitter Lite?

Fairly busy on low end devices due to the virtual scroller

**Comment by comp615** [comment]
Based on the discussion in the whatwg thread, it seems like the heaviest / blocking part is the decoding, though I've not run any tests to actually try that out. I was under the impression that their spec is designed to solve the same problem that the rIC workaround in RNW is trying to solve, namely preventing jank as images decode/paint. Though there may be additional problematic parts (enqueuing all the requests, etc.), I believe this is the core problem RNW's rIC is trying to solve.

Approaching the problem from a different perspective, (or resolving from scratch), in a world where `decode="async"` exists: It seems like RNW should 100% proxy that through to the underlying element. I think we are all agreed there?

So in that world, we'd be talking about RNW providing a workaround for unsupported browsers so that we are respecting user intent when they send the decode param. I think therefore what would make the most sense to me as a developer is:
1) provide the decode param on the Image component as a Web-only option with auto, async and sync. (consistent with the whatwg spec)
2) In cases where the browser does not natively support that option AND async is specified, workaround it with the rIC + timeout combination mentioned above. 

I recognize the behavior is slightly different, (loading vs decoding), but it's as close as you can really get in many cases. 

I suggest that because, as a developer, my expectation would be that by default RNW elements work as closely to the spec, or (where not applicable) to the existing web behavior as possible. As it stands today I, as a user of RNW, don't have control over this potentially controversial behavior so RNW shouldn't do more than I ask of it, as that may be confusing. 

Adding this parameter would let me control it in a way consistent with the browser's spec, and it would let me explicitly choose the behavior. Adding a fallback when async is explicitly specified would mean I no longer need to be worried about browser support, but rather can trust the RNW spec in all cases to approximate my intent.

In short, RNW would be accepting the browser proposal into its own spec (which would supersede existing functionality) as a way of letting users choose.

Sorry for the essay. @sidhu663 do you have any demo code we can talk about / see to ground the discussion more?

**Comment by necolas** [comment]
[Proposed patch](https://github.com/necolas/react-native-web/commit/84da5face4bc8bd7a539b755c21c9476fd11ac46). Using the `decode` API can be figured out separately.

**Comment by necolas** [comment]
MDN docs recommend including a timeout for required work:

> A timeout option is strongly recommended for required work, as otherwise it's possible multiple seconds will elapse before the callback is fired.

**Comment by comp615** [comment]
Timeout is a good start!

>unless it's clear that opt-in to sync decoding has important use cases

I'm not sure how strongly I can advocate for this, but in general I'll say many people debated it at length on the other thread and ended up with the tri-state approach (over just a boolean).

The one thing that did come to mind in my reading of it was that the load event is triggered before decoding when async. Put differently, sync in theory ensures that there is not a secondary flicker/paint. They mention that cases where the user wants to really control everything appearing at once could be important (maybe large image layouts). 

In Twitter Lite, the use case of data saver comes to mind. When a user clicks the preview to load the full image, that image could be sync. Because we swap the images out on the load event, and the loading of the image is the primary focus, we wouldn't want the image to async decode as you'd potentially see preview -> blank el -> full.

Is that reasonable?

**Comment by necolas** [comment]
Yes that seems reasonable when you're working with dom directly. I'll look into how this works in practice and could be used, plus talk to the react native team about it

**Comment by kurtextrem** [comment]
I'm a bit late to the party, but wouldn't replacing rIC with Web Workers work too?
